File: payment-virtualaccount-api/php/virtual_account.php
<?php
error_reporting(E_ALL);
ini_set("display_errors", true);

$path = "http://127.0.0.1/Samples";


$orderId = "virtualaccount-" . time()*1000;
$amount = 50000;
$customerEmail = "customer@email.com";
$customerName = "박토스";
$orderName = "토스 가상계좌 결제";
$bank = "국민";
$validHours = 72;
$virtualAccountCallbackUrl = $path . "/va_callback.php";
$customerMobilePhone = "01000001234";
$useEscrow = false;



$type = "소득공제";
$registrationNumber  = "01000001234";


$secretKey = 'test_ak_ZORzdMaqN3wQd5k6ygr5AkYXQGwy'; 

$url = 'https://api.tosspayments.com/v1/virtual-accounts';

$data = ['orderId' => $orderId, 'amount' => $amount, 'customerEmail' => $customerEmail, 'customerName' => $customerName,
'bank' => $bank, 'validHours' => $validHours,'virtualAccountCallbackUrl' => $virtualAccountCallbackUrl,
'customerMobilePhone' => $customerMobilePhone,'useEscrow' => $useEscrow, 'orderName' => $orderName,
'cashReceipt' => ['type' => $type,'registrationNumber' => $registrationNumber] ];

$credential = base64_encode($secretKey . ':');

$curlHandle = curl_init($url);


curl_setopt_array($curlHandle, [
    CURLOPT_POST => TRUE,
    CURLOPT_RETURNTRANSFER => TRUE,
    CURLOPT_HTTPHEADER => [
        'Authorization: Basic ' . $credential,
        'Content-Type: application/json'
    ],
    CURLOPT_POSTFIELDS => json_encode($data)
]);





$response = curl_exec($curlHandle);

$httpCode = curl_getinfo($curlHandle, CURLINFO_HTTP_CODE);
$isSuccess = $httpCode == 200;
$responseJson = json_decode($response);



?>



<!DOCTYPE html>
<html lang="ko">
<head>
    <title>결제 성공</title>
    <meta http-equiv="x-ua-compatible" content="ie=edge"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"/>
</head>
<body>
<section>
    <?php
    if ($isSuccess) { ?>
        <h1>결제 성공</h1>
        <p>결과 데이터 : <?php echo json_encode($responseJson, JSON_UNESCAPED_UNICODE); ?></p>
        <p>orderName : <?php echo $responseJson->orderName ?></p>
        <p>method : <?php echo $responseJson->method ?></p>
        <p>virtualAccount -> accountNumber : <?php echo $responseJson->virtualAccount->accountNumber ?></p>
        <p>virtualAccount -> bank  : <?php echo $responseJson->virtualAccount->bank  ?></p>
              
        <?php
    } else { ?>
        <h1>결제 실패</h1>
        
        <p>에러메시지 : <?php echo $responseJson->message ?></p>
        <span>에러코드: <?php echo $responseJson->code ?></span>
        <?php
    }
    ?>

</section>
</body>
</html>

File: payment-virtualaccount-api/go/virtual-account.go
package main

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
)

const secretKey string = "test_ak_ZORzdMaqN3wQd5k6ygr5AkYXQGwy"

// VirtualAccount -
type VirtualAccount struct {
	OrderId      string `json:"orderId"`
	OrderName    string `json:"orderName"`
	Amount       string `json:"amount"`
	CustomerName string `json:"customerName"`
	Bank         string `json:"bank"`
}

func basicAuth(username, password string) string {
	auth := username + ":" + password
	return base64.StdEncoding.EncodeToString([]byte(auth))
}

func virtualAccountHandler(w http.ResponseWriter, r *http.Request) {

	fmt.Fprintf(w, "가상계좌 발급 요청 API.\n")

	orderId := "es_test_20220707_001"
	orderName := "토스-슬리퍼 외 2건"
	amount := "200"
	customerName := "프로페서신"
	bank := "국민"

	// Request Parameter json 형태로 셋팅
	va := VirtualAccount{orderId, orderName, amount, customerName, bank}
	pbytes, _ := json.Marshal(va)

	buff := bytes.NewBuffer(pbytes)

	client := &http.Client{
		Transport: nil,
		Jar:       nil,
		Timeout:   0,
	}

	req, err := http.NewRequest("POST", "https://api.tosspayments.com/v1/virtual-accounts", buff)
	req.Header.Add("Authorization", "Basic "+basicAuth(secretKey, ""))
	req.Header.Add("Content-Type", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()

	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(data))
	fmt.Fprintf(w, "\n\n== 가상계좌 발급 요청 응답 결과 ==\n")
	fmt.Fprintf(w, string(data))
}

func main() {
	http.HandleFunc("/", virtualAccountHandler)
	log.Fatal(http.ListenAndServe(":9090", nil))
}
